Add options for Dotty to create semanticDB when Metals is present

tgodzik opened the pull request:
This changes are needed so that user experience for Dotty will be exactly the same as it in Scala 2.

Previously, user would have to add the settings manually.

We only need 2 settings in this case: `-Ysemanticdb` and `-sourceroot`

@jvican Let me know if this makes sense.

olafurpg commented on frontend/src/main/scala/bloop/data/Project.scala:
```diff
@@ -271,18 +273,29 @@ object Project {
+    val options = projectWithRangePositions.scalacOptions
+
     semanticDBPlugin match {
+      case None if isDotty =>
```
nit: instead of guarding in the pattern match, I find it more readable to put the guard inside the body
```suggestion
      case None =>
        if (isDotty) ...
        else ...
```

olafurpg commented on frontend/src/main/scala/bloop/data/Project.scala:
```diff
@@ -271,18 +273,29 @@ object Project {
       (baseOptions ++ baseSemanticdbOptions ++ pluginOption).distinct
     }
 
+    def enableDottySemanticDB(options: List[String]) = {
```
nit: the convention we use is lowercase db for better or worse
```suggestion
    def enableDottySemanticdb(options: List[String]) = {
```

tgodzik commented:
There seems to be an issue running the new tests on JDK 11 :/

```
java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$AppClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
	at xsbt.CompilerClassLoader.computeFixedLoader(CompilerClassLoader.java:113)
	at xsbt.CompilerClassLoader.lambda$fixBridgeLoader$0(CompilerClassLoader.java:98)
	at java.base/java.util.Map.computeIfAbsent(Map.java:1003)
	at xsbt.CompilerClassLoader.fixBridgeLoader(CompilerClassLoader.java:98)
	at xsbt.CompilerInterface.newCompiler(CompilerInterface.java:31)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:252)
	at sbt.internal.inc.AnalyzingCompiler.newCachedCompiler(AnalyzingCompiler.scala:144)
	at sbt.internal.inc.AnalyzingCompiler.newCachedCompiler(AnalyzingCompiler.scala:131)
	at sbt.internal.inc.FreshCompilerCache.apply(CompilerCache.scala:100)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
	at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.compileSources$1(BloopHighLevelCompiler.scala:142)
	at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$7(BloopHighLevelCompiler.scala:160)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$1(BloopHighLevelCompiler.scala:72)
	at bloop.tracing.BraveTracer.trace(BraveTracer.scala:31)
	at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.timed$1(BloopHighLevelCompiler.scala:71)
	at sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$6(BloopHighLevelCompiler.scala:160)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
```

Probably something we need to fix in zinc?

romanowski commented:
I think this line is to blame: https://github.com/lampepfl/dotty/blob/master/sbt-bridge/src/xsbt/CompilerClassLoader.java#L102

jvican commented on frontend/src/test/scala/bloop/bsp/BspMetalsClientSpec.scala:
```diff
@@ -331,6 +331,53 @@ class BspMetalsClientSpec(
+      )
+      loadBspState(workspace, projects, logger) { state =>
+        val compiledState = state.compile(`A`).toTestState
+        assert(compiledState.status == ExitStatus.Ok)
```
Let's try to add some of the other checks we have in `CompileSpec` for simple hello world compilations in Scala 2. You can find these checks at the beginning of the `CompileSpec` definition.

tgodzik replied:
Should we create a separate CompileSpec for Scala 3? Or do you mean to add them here?

tgodzik replied:
Or just he asserts from the test cases there?

jvican commented on frontend/src/test/scala/bloop/bsp/BspMetalsClientSpec.scala:
```diff
@@ -331,6 +331,53 @@ class BspMetalsClientSpec(
+      loadBspState(workspace, projects, logger) { state =>
+        val compiledState = state.compile(`A`).toTestState
+        assert(compiledState.status == ExitStatus.Ok)
+        assertSemanticdbFileFor("Foo.scala", compiledState, isPresent = false)
```
Same here as before :smile:

jvican commented on frontend/src/test/scala/bloop/bsp/BspMetalsClientSpec.scala:
```diff
@@ -357,13 +404,16 @@ class BspMetalsClientSpec(
-      classesDir.resolve(s"META-INF/semanticdb/A/src/$sourcePath.semanticdb")
-    )
+    val file = classesDir.resolve(s"META-INF/semanticdb/A/src/$sourcePath.semanticdb")
+    if (isPresent) assertIsFile(file) else assertNotFile(file)
```
Nit: I propose we introduce a new method `assertNotSemanticdbFileFor` that checks the non-existence of a semanticdb file, just as we do with the combo `assertIsFile` and `aasertNotFile`.

tgodzik commented:
@jvican Seems we will not be able to run Dotty compilation on JDK 11 due to the [issue](https://github.com/lampepfl/dotty/blob/master/sbt-bridge/src/xsbt/CompilerClassLoader.java#L102) @romanowski found :/ 

So either we don't run tests for now or we have to fix the issue in Dotty itself.

alexarchambault commented:
FYI, if you're not familiar with that JDK11 issue, there's [a fix for a similar issue in metabrowse](https://github.com/scalameta/metabrowse/blob/c405b5be580e4925252390763955ad9eb9e266f8/metabrowse-tests/src/test/scala/metabrowse/tests/JsonSuite.scala#L26-L35). It basically amounts to assume the `getURLs` of a `jdk.internal.loader.ClassLoaders$AppClassLoader` is `System.getProperty("java.class.path").split(File.pathSeparator)`. It's described [there](https://blog.codefx.org/java/java-11-migration-guide/#Casting-To-URL-Class-Loader) too.

tgodzik commented:
@alexarchambault Thanks! I did find the issue, but still not sure exactly where to fix it.

Got something working here: https://github.com/scalacenter/zinc/compare/1.x-bloop...tgodzik:fix-classpath

tgodzik commented:
@jvican I updated Dotty version to a nightly with a fix, otherwise the tests will not work.

I also added a couple of checks to the test case, but couldn't find too many more. Some workflows are being tested there in `CompileSpec`, but I don't believe it's a place to add them here in `BspMetalsClientSpec`. What checks did you mean exactly?

jvican commented on frontend/src/test/scala/bloop/bsp/BspMetalsClientSpec.scala:
```diff
@@ -331,6 +331,57 @@ class BspMetalsClientSpec(
+      loadBspState(workspace, projects, logger) { state =>
+        val compiledState = state.compile(`A`).toTestState
+        assertExitStatus(compiledState, ExitStatus.Ok)
+        assertValidCompilationState(compiledState, projects)
```
Re my previous comments, these two assertions are enough :smile: It's important that we assert the compilation state is valid and it's a successful compilation because there are lots of invariants that need to be met for a compilation to be correct. 👍
